pkgs/development/coq-modules/hierarchy-builder/default.nix

```diff
@@ -0,0 +1,43 @@
+{ stdenv, fetchFromGitHub, which, coq, coq-elpi }:
+
+let
+  versions = {
+      "0.9.0" =  {
+        rev = "v0.9.0";
+        sha256 = "1yss9f732r7bjaswgn46vd1rr3688ir0vz37wxkmy598xhrnd2ak";
+      };
+  };
+  version = x: versions.${x} // {version = x;};
+  params = {
+   "8.10" = version "0.9.0";
+   "8.11" = version "0.9.0";
+  };
+  param = params.${coq.coq-version};
+in
+
+stdenv.mkDerivation rec {
+  name = "coq${coq.coq-version}-hierarchy-builder-${param.version}";
+
+  src = fetchFromGitHub {
+    owner = "math-comp";
+    repo = "hierarchy-builder";
+    inherit (param) rev sha256;
+  };
+
+  nativeBuildInputs = [ which ];
+  buildInputs = [ coq coq.ocaml coq.ocamlPackages.elpi coq-elpi ];
```

Comment:
coq-elpi should be propagated.

In the current state, running “From HB Require structures.” fails.

Reply:
How do you test this quickly?

Reply:
I never understand what should be propagated, e.g. shouldn't coq be propagated?

Reply:
To test: `nix-shell -p coqPackages_8_10.hierarchy-builder coqPackages_8_10.coq` and then run `coqc` on a test case (e.g., https://raw.githubusercontent.com/math-comp/hierarchy-builder/master/demo1/hierarchy_0.v).

Should `coq` be propagated? No. It is likely to use a different version of Coq to build the package and to use it. In the first case, for instance, we don’t need CoqIDE, the manual etc.

Reply:
> To test: `nix-shell -p coqPackages_8_10.hierarchy-builder coqPackages_8_10.coq`

That's what I tried and it does not work for me:
```
$ nix-shell -p coqPackages_8_10.hierarchy-builder                                                                                                   (hierarchy-builder|✔)
error: attribute 'hierarchy-builder' missing, at (string):1:94
(use '--show-trace' to show detailed location information)
```

Reply:
Can be fixed by setting the `NIX_PATH` environment variable to the path of the directory containing `nixpkgs`.

Reply:
I usually do it on the command-line using the `-I` option.

Reply:
@vbgl  @Zimmi48
> Should `coq` be propagated? No. It is likely to use a different version of Coq to build the package and to use it. In the first case, for instance, we don’t need CoqIDE, the manual etc.

I thought I'd understand this later, but I just had a discussion with @gares and I still don't...
How could a different version of coq be compatible with hierarchy builder since the compiled vo are likely to be compatible only with the coq version they were compiled with?

Plus it would be really nice that hierarchy builder is available through `nix-shell -p coqPackages_8_11.hierarchy-builder`

Reply:
Your CI machine that builds HB does not have CoqIDE nor the Coq manual (to avoid wasting brandwidth). But you want to use it with your local version of Coq that has all these extra features.

This is one possible use-case. More generally you don’t want to propagate a compiler.

Reply:
@vbgl can you explain why the dependency on Coq is not "propagated".
I mean, HB without Coq serves no purpose. I don't get what you mean by "It is likely to use a different version of Coq to build the package and to use it", as far as I now there is no guarantee that a .vo compiled with Coq 8.x.0 works with 8.x.1 (assuming they are compiled with the same ocaml compiler). It would be nice, but I don't think we try to ensure that.

Reply:
@vbgl isn't my local version of coq always the same? Aren't all possible customization (that do not alter the compilation and thus the vos) external packages (e.g. coqide, coq-elpi, etc) rather than overrides of coq?

Reply:
I had in mind something like that, where the Coq package you use is not the same as the Coq package for building third-party libraries:

~~~
{ pkgs ? import <nixpkgs> {}
}:

with pkgs;

let
 # Coq for building libraries
 coqTiny = (coq_8_10.override { buildIde = false; csdp = null; })
  .overrideAttrs (o: { enableParallelBuilding = true; });
 coqPackages = mkCoqPackages coqTiny;
 # Coq for the end user
 coqFat = coq_8_10;
in

stdenv.mkDerivation {
name = "test";

buildInputs = [ coqFat coqPackages.ssreflect ];
}
~~~

The `.vo` files are identical.

However, this does not work for OCaml plug-ins: some CRC check fail at load time:

> Error: while loading
> /nix/store/an5bgv69msbi4vki9m5lck13rayakdgj-coq8.10-elpi-1.3.0/lib/coq/8.10/user-contrib/elpi/elpi_plugin.cmxs:
implementation mismatch on Vernacextend

I have no clue what causes the differences in these OCaml libraries…

Reply:
Maybe I'm just pessimistic, but if it is not enforced upstream the fact that .vo files works is just a miracle and you should not rely on it.

For the cmxs, the thing is that OCaml does not really has an ABI to check, they take the md5 of the file. So any change to any Coq module (like Vernacextend, and even if the mli does not change) makes object files compiled against the old version not load/link against the new version. If you compile everything with -opaque (which is not the default, for example, in release mode for dune projects) then you get cross-module inlining, so it is really unsound to load code compiled against an old version of a module (you may have in the object file code belonging to the old version of the other module).

Hope it helps

Reply:
Let’s start from two identical configurations (same OCaml, same snapshot of the Coq source code). In the first one you run `./configure -coqide no` and in the second, you run `./configure -coqide opt`.

Are you saying that the fact that using the .vo from the first configuration in the second one does work is a miracle?

I’m still puzzled to get different .cmx files.

Reply:
> Are you saying that the fact that using the .vo from the first configuration in the second one does work is a miracle?

No, in that case I believe they should be the same looking at today's sources. But even a single `#IFDEF COQIDE` could break this.

If it does not work today, then I believe there is something very fishy going on.

I'm not familiar with nix. Do you have 1 source - 1 binary as in opam?

Reply:
Yeah, it would be nice to have part of the standard library dedicated to CoqIDE users! (just kidding).

> Do you have 1 source - 1 binary

What do you mean?

Reply:
That I don't know why you would compile coq twice. I debian the output of the compilation of 1 package can generate many packages, so coq & coqide are compiled once and later split. Here you seem to suggest that you compile coq twice.

Reply:
@gares in nix, it's 1 derivation -> 1 package, a derivation is the data of "the source + the dependencies + the commands that are run to fix the sources, build, fixup stuff and install."
Hence if you tune some compiling options, you might create a different coq binary but you definitely create a new nix coq package.